lms/services/application_instance_getter.py

```diff
@@ -0,0 +1,117 @@
+        :rtype: str or ``None``
+        """
+        return self._get(consumer_key).developer_key
+
+    def developer_secret(self, consumer_key):
+        """
+        Return the Canvas developer secret for the given consumer_key, or None.
+
+        :arg consumer_key: the consumer key to search for
+        :type consumer_key: str
+
+        :raise ConsumerKeyError: if the consumer key isn't in the database
+
+        :return: the matching Canvas API developer secret or ``None``
+        :rtype: str or ``None``
+        """
+        application_instance = self._get(consumer_key)
+
+        developer_secret = application_instance.developer_secret
+        if developer_secret is None:
+            return None
+
+        cipher = AES.new(
+            self._aes_secret, AES.MODE_CFB, application_instance.aes_cipher_iv
+        )
+        return cipher.decrypt(developer_secret)
+
+    def lms_url(self, consumer_key):
+        """
+        Return the LMS URL for the given LTI consumer_key.
+
+        :arg consumer_key: the consumer key to search for
+        :type consumer_key: str
+
+        :raise ConsumerKeyError: if the consumer key isn't in the database
+
+        :return: the matching LMS URL
+        :rtype: str
+        """
+        return self._get(consumer_key).lms_url
+
+    def provisioning(self, consumer_key):
+        """
+        Return ``True`` if provisioning is enabled for the given consumer key.
+
+        Return ``True`` if the provisioning feature is enabled for the given
+        consumer key, ``False`` otherwise.
+
+        :arg consumer_key: the consumer key to search for
+        :type consumer_key: str
+        """
+        try:
+            provisioning = self._get(consumer_key).provisioning
+        except ConsumerKeyError:
+            provisioning = False
+        return provisioning
+
+    def shared_secret(self, consumer_key):
+        """
+        Return the LTI/OAuth 1 shared secret for the given LTI consumer_key.
+
+        This is called the ``oauth_consumer_secret`` in the OAuth 1.0 spec.
+
+        :arg consumer_key: the consumer key to search for
+        :type consumer_key: str
+
+        :raise ConsumerKeyError: if the consumer key isn't in the database
+
+        :return: the matching shared secret
+        :rtype: str
+        """
+        return self._get(consumer_key).shared_secret
+
+    def _get(self, consumer_key):
+        """
+        Return the ApplicationInstance with the given consumer_key or ``None``.
+
+        :arg consumer_key: the consumer key to search for
+        :type consumer_key: str
+
+        :raise ConsumerKeyError: if the consumer key isn't in the database
+
+        :return: the matching ApplicationInstance or ``None``
+        :rtype: :cls:`lms.models.ApplicationInstance` or ``None``
+        """
+        try:
+            return (
+                self._db.query(ApplicationInstance)
+                .filter(ApplicationInstance.consumer_key == consumer_key)
```

Comment:
Does the result of this query get cached by SQLAlchemy? If not, then some of the code using this service, if it needs to access multiple properties of the install, is going to execute multiple requests against the DB when one would suffice.

Reply:
I haven't checked but I don't think it gets cached. There might be situations with concurrent requests where caching it could introduce problems. But most likely we just want to cache these query results. But I'd like to leave it for a later PR.

I'll be adding an index to `consumer` in a later PR too

Reply:
> There might be situations with concurrent requests where caching it could introduce problems.

Can you give me an example of what you mean? I'm referring to the kind of caching that some of h's services do with the `lru_cache_in_transaction` helper, where each service instance (and therefore each request) gets its own caches which expire at the end of the request.

Reply:
I'm thinking of the kind of duplicate creation problems we've been seeing in h:

1. Request A checks whether a given `ApplicationInstance` exists or not
2. Concurrent request B creates or deletes the `ApplicationInstance`
3. Request A tries to take action based on previous result for whether `ApplicationInstance` exists or not

Does caching query results in Python introduce further possible problems here?

But in the case of `ApplicationInstance`s and what the app currently does with them, I don't think there's a problem

Reply:
> Does caching query results in Python introduce further possible problems here?

In your example, there is a hazard even if no caching is used:

1. Request A executes SQL query to check for existence of a row, the DB query itself returns false but control has not yet returned to Python
2. Request B inserts the row that was queried
3. Control returns to the Python code in Request A which now executes logic with an out-of-date view of whether the object exists or not

Depending on what the logic does, that may or may not be a problem. Either the problem needs to not matter enough (eg. the next time the user refreshes the page, the status is correct), or it needs to be prevented (eg. by making conflicting changes impossible), or it needs to be handled gracefully (eg. by handling conflicts when a subsequent transaction commit fails).

It is hard to say that lookup caching within a request could _never under any circumstances_ cause any additional complications, I don't think that it fundamentally adds extra complexity here.

Reply:
btw. The kind of caching that I'm thinking of is similar to what services in h do with eg. `lru_cache_in_transaction` or just private `_foo_cache` fields on service instances.